### ISSUE TITLE
Duplicate CSS using @import and require()

### ISSUE THREAD

**Issue by mikelew42** [webpack-contrib/less-loader#141]
* If you want access to LESS variables, you have to use `@import`.  LESS variables are not available from one `require()` to the next (see "Simple Example" section, at the very end).
* This means each `require()` creates its own LESS compiling context, and if you want to use variables between `.less` files, *you have to use a single `require()` statement, and use `@import` to pull in the files*.
* This is very counter-intuitive to the traditional `npm` system, where we can define a dependency and `require()` it, and it `requires()` its deps, (in this case, LESS styles).  Instead of using `require()` for each component's `.less`, we have to edit our global `.less` bundle, and manually `@import` everything separately from the JavaScript.

This isn't impossible, just a little extra work.  And, all modules would have to follow this pattern (of not `requiring()` their own `.less`, and instead allowing the end user to pull it in with `@import`).  If you're the author of all your components, that's possible.  Anyway, it's not a great solution, but here's how that might work:

## Solution A:  Manage LESS and JS separately

Maintain a single `global.less` file, and make sure to only require it once.  In `global.less`, use `@import` to pull in all the individual `.less` files, where ever they may be (local or in `node_modules`).  For example,  `@import "~/path/to/nav.less"`.

Also, in order to render the nav component's HTML, we'd need to `require()` the necessary JS: `require("path/to/nav")`, wherever that would normally happen.

**How it should work** 
 
We `require("path/to/nav");` which then `requires("./nav.less")`, which can than `@import` mixins that have already been `required()` and `@imported`, without creating duplicate CSS.

TL;DR - That's basically it... The following might help someone, so I decided to leave it:

## Solution B:  Keep vars/mixins in separate files without any CSS output

LESS mixins and variables aren't output to CSS anyway - so if we isolate all our LESS vars and mixins in a separate file, it's OK if they are improperly `required()` and `@imported` a million times.  The mixins and vars get stripped.

The problem here, is that most of my mixins aren't *abstract* - they are real CSS rules, that I also use as mixins.  For example:

I add a CSS class to an element as a mixin/override

```
nav.addClass("large-padding"); // we need the real CSS rule defined
```

Or, I use it as a LESS mixin:

```less
.nav {
    .large-padding;
}  
```

**UPDATE: You can have `.large-padding(){ ... }` *and* `.large-padding { .large-padding(); }` defined simultaneously.**

I had originally described using an alternate `.large_padding()` mixin, to avoid namespace clash, to do something similar - but apparently they are namespaced separately, so we can have both.

This means that if we want the `.large-padding` css rule defined, and available as a mixin, we just need to do both strategically:

`mixins.less`:

```less
.my-class(...) {
    ...
}
```

And then also define it in your `global.less`:

```less
@import "./mixins.less";
.my-class {
    .my-class(); // use defaults or whatever
}
```

And then elsewhere, we can `@import "~pkg/mixins.less"`, and use all the mixins.

## Simple example: `@import` vs `require()`:

Consider the following setup:

### `one.less`:

```less
.something {
    /* css rules here */
}
```

### `two.less`:

```less
.class {
    .something;
}
```

### and then in our JavaScript:

```
require("./one.less");
require("./two.less");
```

### This throws an error

The variables defined in `one.less` aren't available in `two.less`.  It seems each `require()` creates a separate LESS compiling scope.

> Note: I'm using the extract text webpack plugin - not sure if that changes any of this

### `@import`, however, works

In a `root.less` file:

```less
@import "one.less";
@import "two.less";
```

and then w/ JavaScript:

```
require("./root.less");
```

This works - there is a single `require()` statement, and the two `.less` files are pulled in with `@import` statements, so there is a single LESS compiling scope, and the variables are available from `one.less` to `two.less`.

The problem is that if you `require("./root.less")` in your bundle (which pulls in `one.less` and `two.less`), and then `require("./another.less")` in which you need access to vars defined in `one.less` or `two.less`, you need to `@import` them (even though they've already been `require()`.  This creates duplicate CSS in your final ouput.

**Comment by mikelew42** [comment]
Probably a similar issue:
https://github.com/webpack/less-loader/issues/100

**Comment by mikelew42** [comment]
Conclusion:  I want to author/publish/`require()` npm modules that can extend other npm modules.  This is possible with JavaScript.  To do this with LESS means we need to be able to `@import`

**Comment by mikelew42** [comment]
[This article](https://www.bensmithett.com/smarter-css-builds-with-webpack/) discusses this topic, for additional reading.

I'm still trying to figure out how I'm going to proceed...

**Comment by mikelew42** [comment]
Update:  I'm going to try Solution B described above.  Read the update I added in the original post... Ask me if you're interested and can't make sense of it.

**Comment by mikelew42** [comment]
Update, I stumbled upon [this SO post](http://stackoverflow.com/questions/14453737/reference-a-class-mixin-without-completely-importing-the-less-file?rq=1) and realize that this ability to @import as a reference only solves some of my challenges.

Now, my approach will be something like this:

    /components/item/
        item.js
            require("./item.less");
    
    /components/item2/
        item2.js
            require("./item/item.js");  // this adds item.less to the CSS output, if not already required()
            require("./item2.less");
        item2.less
            @import (reference) "../item/item.less"; // provides all LESS vars/mixins

This would normally cause duplicate CSS output, if you didn't use the (reference) feature.  The (reference) flag just tells LESS to import all the vars/mixins with 0 output.  

I have yet to test this, but thought I'd keep a log of my journey.

**Comment by linkerx** [comment]
@mikelew42 last post is a great solution and to me is a very clean one.
this way can add separated less for each component and @import variables and mixins but without the duplicated output.

thanks!

**Comment by michael-ciniawsky** [comment]
Duplicate of #7
